test/helpers/sockets.rb

```diff
@@ -0,0 +1,171 @@
+        begin
+          # part = sysread RESP_READ_LEN
+          part = read_nonblock(RESP_READ_LEN, exception: false)
+          case part
+          when String
+            unless content_length || chunked
+              chunked ||= part.include? "\r\nTransfer-Encoding: chunked\r\n"
+              content_length = (t = part[/^Content-Length: (\d+)/i , 1]) ? t.to_i : nil
+            end
+
+            response << part
+            hdrs, body = response.split RESP_SPLIT, 2
+            unless body.nil?
+              # below could be simplified, but allows for debugging...
+              ret =
+                if content_length
+                  # STDOUT.puts "body.bytesize #{body.bytesize} content length #{content_length}"
+                  body.bytesize == content_length
+                elsif chunked
+                  # STDOUT.puts "#{body.bytesize} chunked"
+                  body.end_with? "\r\n0\r\n\r\n"
+                elsif !hdrs.empty? && !body.empty?
+                  true
+                else
+                  false
+                end
+              if ret
+                @connection_close = hdrs.include? "\nConnection: close"
+                return [hdrs, body]
+              end
+            end
+            sleep 0.000_1
+          when :wait_readable, :wait_writable # :wait_writable for ssl
+            sleep 0.000_2
+          when nil
+            @connection_close = true
+            raise EOFError
+          end
+        end
+      end
+      raise Timeout::Error, 'Client Read Timeout'
+    end
+
+    # Reads the raw response as string
+    # @!macro tout
+    # @return <String>
+    def read_raw(timeout = nil)
+      timeout ||= RESP_READ_TIMEOUT
+      response = +''
+      timeout_end = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
+      while to_io.wait_readable timeout
+        timeout = timeout_end - Process.clock_gettime(Process::CLOCK_MONOTONIC)
+
+        part = read_nonblock(RESP_READ_LEN, exception: false)
+        case part
+        when String
+          response << part
+          sleep 0.000_1
+        when :wait_readable, :wait_writable # :wait_writable for ssl
+          sleep 0.000_2
+        when nil
+          return response
+        end
+      end
+      raise Timeout::Error, 'Client Read Timeout'
+    end
+  end
+
+  unless Object.const_defined?(:Puma) && ::Puma.const_defined?(:HAS_SSL) && !Puma::HAS_SSL
+    require 'openssl'
+    # The SSLSocket class used by the TestPuma framework.  The `SktPrepend` module
+    # is prepended.  The socket is opened with parameters set by `bind_type`.
+    class SktSSL < ::OpenSSL::SSL::SSLSocket
+      prepend SktPrepend
+    end
+  end
+
+  # The TCPSocket class used by the TestPuma framework.  The `SktPrepend` module
+  # is prepended.  The socket is opened with parameters set by `bind_type`.
+  class SktTCP < ::TCPSocket
+    prepend SktPrepend
+  end
+
+  if Object.const_defined? :UNIXSocket
```

Comment:
What does moving this logic into a module which is prepended to a custom socket class get us that we didn't have before? I guess I'm not thrilled by having to change the method calls everywhere, and for newbies to Puma contributing to have to learn that we have a special socket class that we use.

Reply:
> What does moving this logic into a module which is prepended to a custom socket class get us that we didn't have before?

Correct handling of sockets.  Using read, write, gets, etc uses buffering in Ruby, which can be preempted by other code.  As I saw recently, a test took 20 seconds with read, but less than a second with sysread/read_nonblock.

Also, this is the start of the conversion.  Two additional helper files handle creation of in-proc servers and popen servers.  Each has a method (`startServer` and `startPuma`) with similar setup.  Bind can be `tcp`, tcp4`, `tcp6`, the same for ssl, along with `aunix` and `unix`.  To create a client socket, a `connect` method is used, and it knows what kind of socket to create.

The server helper files and the sockets file are aware of each other.  In this PR I removed all the methods from sockets.rb that are dependent on the server helper files.

Additionally, we have issues with sockets not being closed, unix socket files remaining, etc.  If one looks over the current test suite, client sockets are created in different ways, same for servers.  IOW, there is probably a lot duplicated 'helper' code.

I think most people write tests by looking at other similar tests.  We can also doc the new test DSL methods, I thought placing them in a `TestPuma` namespace would allow them to be included in the docs without much distraction.  If the additional methods clutter up the methods list, we can just doc them with md files.

Lastly, test times
```
Ubuntu test_puma_server.rb
        This PR      master
serial    26.0       86.0
parallel   3.1       40.9 
```

**Off-topic**
Regardless of what we do with this, maybe it's time to divide up some of the test files?  It varies from file to file, but TestPumaServer has 96 tests...